### PR TITLE
pyppeteer

### PR DESCRIPTION
**Pull request by kiwi0fruit**

### CODE REVIEW

**Comment by kiwi0fruit** [comment]
@conda-forge-admin @conda-forge/help-python please review.

File: recipes/pyppeteer/meta.yaml
Comment (igortg) [inline recipes/pyppeteer/meta.yaml]: `noarch: python` should be used for Python 2 and 3 compatible libraries

File: recipes/pyppeteer/meta.yaml
Comment (igortg) [inline recipes/pyppeteer/meta.yaml]: The lib seems to be compatible with py35

File: recipes/pyppeteer/setup.patch
Comment (igortg) [inline recipes/pyppeteer/setup.patch]: Why this patch is needed?

File: recipes/pyppeteer/meta.yaml
Comment (kiwi0fruit) [reply recipes/pyppeteer/meta.yaml]: The noarch wiki page says othewise. Noarch is about win,  macos, linux builds. The fact that it can unify python versions too is just a bonus.

File: recipes/pyppeteer/meta.yaml
Comment (kiwi0fruit) [reply recipes/pyppeteer/meta.yaml]: The py35 support is stated to experimental and needs one extra dependence: py-backwards, that is not on conda-forge. Hence patch.

File: recipes/pyppeteer/meta.yaml
Comment (kiwi0fruit) [reply recipes/pyppeteer/meta.yaml]: https://conda-forge.org/docs/meta.html#building-noarch-packages

**Comment by kiwi0fruit** [comment]
@igortg what about this PR?

**Comment by kiwi0fruit** [comment]
@beckermr Please take a look at this PR too.

**Comment by kiwi0fruit** [comment]
I have concerns that pyppeteer downloads chromium build to the appdata/userdata by default. I find conda packaging Chromium a waste of time as pyppeteer already does it. But maybe better to patch it for secure downloading: https://github.com/miyakogi/pyppeteer/pull/189/files

**Comment by kiwi0fruit** [comment]
Or maybe even set folder for chromium downloading to roor conda env or something?

File: recipes/pyppeteer/meta.yaml
Comment (beckermr) [reply recipes/pyppeteer/meta.yaml]: We don't ship py35 on conda-forge, so this discussion is moot.

Comment (beckermr) [reply recipes/pyppeteer/meta.yaml]: That being said, we do need the patch so it builds.

File: recipes/pyppeteer/setup.patch
Comment (beckermr) [reply recipes/pyppeteer/setup.patch]: to get it to build, right @kiwi0fruit ?

File: recipes/pyppeteer/meta.yaml
Comment (beckermr) [reply recipes/pyppeteer/meta.yaml]: actually, maybe we don't need the patch. We should try it without.

File: recipes/pyppeteer/setup.patch
Comment (kiwi0fruit) [reply recipes/pyppeteer/setup.patch]: Not really. It was more a preventive measure so I would predict how it behaves and make sure it won't give bugs.

Comment (beckermr) [reply recipes/pyppeteer/setup.patch]: Ahhhh ok. I'd say we remove it. I don't think it is doing that much to make things safer and probably creates more maintenance.

**Comment by kiwi0fruit** [comment]
if we are to use conda packaged chromium this most likely to change Pyppeteer dir for storing chromium. but then we should also make sure that Pyppeteer doesn't try to install to that folder when user changes chromium revision via env var...

File: recipes/pyppeteer/meta.yaml
Comment (beckermr) [reply recipes/pyppeteer/meta.yaml]: Conda will not try and build python 3.5, so removing the patch is safe.

**Comment by kiwi0fruit** [comment]
By the way: I wasn't able to find chromium in defaults and conda-forge (but I was on the phone then as I am now) Are you sure that it is in the defaults?

**Comment by kiwi0fruit** [comment]
@beckermr I've checked and it seems like there is no chomium in defaults and conda forge. And packing it is another complex task. So I suggest to let pyppeteer download chromium.

There's only question: to what dir download? Should we go with default in appdata/userdata? Or change to other more conda self-contained?

**Comment by beckermr** [comment]
I think we have to package it unless we can vendor it in such a way as to not interfere with stuff outside the conda env. There is also the question of licensing. The final binary needs to be able to legally ship chromium.

**Comment by kiwi0fruit** [comment]
@beckermr chromium is free software as far as I know so it won't be a problem. Why not let pyppeteer install chromium inside conda environment? At least for now. For example pip does this very thing. We install pip conda package and it installs not-linked packages. So let pyppeteer do the same thing. The only thing to deside: to what subfolder should it install?

What do you think about going this way?

**Comment by beckermr** [comment]
My worry is that if another package wants to do the same, we get a conflict. In general separating out the deps is much better. 

@mingwandroid started a chromium recipe here: https://github.com/AnacondaRecipes/aggregate/tree/master/chromium-feedstock/recipe

Is this not active on the anaconda channels?

**Comment by beckermr** [comment]
LOL

```
name: chromium
# Update this from the file:
# https://chromium.woolyss.com/download/
# https://www.googleapis.com/download/storage/v1/b/chromium-browser-snapshots/o/Mac%2FLAST_CHANGE?generation=1519904483346073&alt=media
version: 540119  # [win]
version: 540120  # [osx]

https://www.googleapis.com/download/storage/v1/b/chromium-browser-snapshots/o/Mac%2F540124%2Fchrome-mac.zip?generation=1519904463732592&alt=media
```

OK.

**Comment by beckermr** [comment]
We need guidance from core before we can proceed. Sorry for causing trouble here @kiwi0fruit. 

@conda-forge/core, can you weight in here?

**Comment by kiwi0fruit** [comment]
@beckermr simple separation by specially named subfolder should be enough. there is also ability to override pyppeteer download folder via env vars. it can mess things (just as much as installing pip to userdata that can also be enabled). If we remove it woudn't stray too far from original package? Isn't it a bad thing? So I guess better have isolated defaults but not change the original package much.

Why woudn't for example `<conda_env>/pyppeteer_chromium` subfolder be OK for now? I guess no other module would mess with it. I already have `<conda_env>\node_modules\npm\node_modules` that also has additionally downloaded modules. So I don't see why it's a bad thing.

**Comment by beckermr** [comment]
Right yeah. This approach can be workable! It is simply that a big dep like involves more than this package and so we have to think about the whole ecosystem. This is why we are waiting for the core maintainers. Again I am sorry for the delay on this.

**Comment by kiwi0fruit** [comment]
That's OK in such situaltions I guess. But I don't really see this as a big dep. It needs space, yeah. But it's only pyppeteer internals and so it's not of other packages business.

**Comment by kiwi0fruit** [comment]
@ocefpaf Could you please take look here.

File: recipes/pyppeteer/meta.yaml
Comment (scopatz) [inline recipes/pyppeteer/meta.yaml]: Please remove extra comments

**Comment by scopatz** [comment]
This looks OK to me overall

**Comment by beckermr** [comment]
Just to be really clear @scopatz, we are ok with packages vendoring chromium?

File: recipes/pyppeteer/meta.yaml
Comment (kiwi0fruit) [reply recipes/pyppeteer/meta.yaml]: Done.

**Comment by kiwi0fruit** [comment]
@beckermr I guess OK really meant OK ;) Let's merge?

**Comment by scopatz** [comment]
> we are ok with packages vendoring chromium?

Ahh no not really.  We should have a separate chrome/chromium package that this depends on.

**Comment by beckermr** [comment]
Looks like not @kiwi0fruit. I am very sorry for the delay. We need to maintain the eco system and so this will have to be blocked. I am very happy to contribute my own dev to time to getting chromium in conda so we can do this quickly. If you start a PR from your fork, then I can make commits to it since I am a maintainer here. Does that sound ok?

**Comment by kiwi0fruit** [comment]
This would also mean changing Pyppeteer so that it can work with conda packaged chromium. This would mess with allowing the user change chromium revision via env var. And overall Pyppeteer behavior that has a Pyppeteer home dir that is can write as he pleases. And so on. Who would fix that?... I can make a blank chromium recipe maybe from the Pandoc recipe...

**Comment by beckermr** [comment]
We can ship a patch to pyppeteer with the conda package to make it use the packaged chromium. We could also submit a PR upstream to make the patching easier. Anything else is fine by me. 

This kind of issue is extremely common and conda-forge in general maintains a fair number of patches to do things like this.

**Comment by isuruf** [comment]
Looks like pyppeteer is not vendoring chromium when building the recipe, but it installs the package at runtime and the user has the ability to select which chromium version to use. Changing this to a conda packaged version is not going to work IMO. How would the user be able to select the chromium version to use using pypperteer's way?

**Comment by kiwi0fruit** [comment]
I guess upsteam at the moment is not responding... I wonder why this happens this way instead of publishing as it is and then fixing to the standard? I works now and doesn't mess with anything.

**Comment by kiwi0fruit** [comment]
@isuruf Yes, it behaves like p*u*ppeteer in this this regard. And I guess this feature is needed for some people. And it's actually a part of pyppeteer desing. I'm against changing pypeteer to something that's no longer pyppeteer.

**Comment by kiwi0fruit** [comment]
@isuruf By the way: during this long conversation I made a patch that relocated folder to where pyppeteer downloads chromium from appdata/userdata to special folder in the environment root. Is it needed or may be the original design is also OK and doesn't need patching?

**Comment by beckermr** [comment]
@kiwi0fruit @isuruf Thanks for the expertise and help! I am ok merging as is then. Chromium is closer to a CDT right now I suppose. 

Thanks again for the patience @kiwi0fruit! Getting this all right on volunteered time is sometimes a very slow process. 

Are we all good?

**Comment by kiwi0fruit** [comment]
@beckermr I'd wait for @isuruf  response about relocating of the dir. If it's not necessary that I would make a smaller patch.

What do you think about dir where pyppeteer downloads chromium? Should it be in appdata/userdata (as it's by design) or we should move it to the special folder in the environment dir?

**Comment by kiwi0fruit** [comment]
@beckermr I guess it's better to relocate to special folder in the env dir so that it behaves like node js. If some arguments against that would appear then we can always revert it back to default.

So it's the current PR. So I ask to pull it. Tests are fine.

**Comment by kiwi0fruit** [comment]
Hooray!

**Review by Harouna1972 — APPROVED** [review]

**Pull request merged**
